### PR TITLE
Improved flexibility of the external API surface.

### PR DESCRIPTION
**Pull request by tacosontitan**
# TL;DR

This PR improves the flexibility of the external API surface by adding abstractions.

## Menu Builder

The primary method of building menus is through an abstraction called `IMenuBuilder` which is created through static methods on the `Menu` class, similar to creating a host builder in hosted .NET applications:

```cs
Menu.CreateDefaultBuilder()
    .SetTitle("Hangman (1.0.0)")
    .SetSelector("[x]")
    .SetPrefix("[ ]")
    .AddOption<PlayOption>()
    .AddOption<CreditsOption>()
    .AddOption(text: "Exit", action: static () => Environment.Exit(0))
    .Build()
    .Run();
```

## Menu Renderer

Menu renderers are the primary driver for how menu elements end up on the screen. Currently, things are limited to a console application, but why stop there? The basics of a renderer are the selector, prefix, theme, and a method to render a specified menu with.

## Elements

Menus are composed with `IMenuElement` instances which could be a basic `MenuTitle` or a fully defined `IMenuOption` implementation. This allows the menu to contain a simple collection of elements to render at any given time.

***NOTE**: The element collection for a menu is now immutable once the menu is created, so you can't run the menu and add another option.*

## Themes

Themes enable consumers to define their look and feel once and let the renderer handle the rest. Currently only the default theme is defined, but it would be ideal to define dark and light mode themes using common combinations of foreground and background colors.

## Next Steps

The new abstractions and their implementations do not currently account for a few scenarios that were previously covered.

### Multiple Selections

This part is now simplified but will take a bit of reworking to the theme extensions and `AnsiConsoleColor` to ensure things still behave as expected. I don't think an entirely different `IMenu` implementation is needed to account for this as it could just be a method on the builder and the default menu implementation could simply account for it:

```cs
Menu.CreateDefaultBuilder()
    .EnableMultipleSelection()
    ...;
```

This way, any builder added to the project in the future will automatically support single and multiple selection.

### Input Management

The input management doesn't currently account for shortcuts as the previous implementations did. This is simply because the new implementations do not have an abstraction to handle this with yet. My recommendation is to create abstractions and implementations to handle input management in a simple way that allows consumers to override and change parts out if needed.

Additionally, the current implementation doesn't account for disabled elements when navigating or invoking. That will also need to be addressed.

### Per Element Customization

Previous implementations would allow consumers to specify customizations on a per-element basis. The new implementations do not permit this yet, but it wouldn't be overly complicated to add. The decision needs to be made if that's considered a critical element for the package's offerings or not.

### CODE REVIEW

**Comment by tacosontitan** [comment]
This is a draft so that you can begin reviewing changes while I'm ironing out some known issues.
